pkg/internal/handler.go

```diff
@@ -0,0 +1,131 @@
+package internal
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/giantswarm/conditions/pkg/conditions"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	capi "sigs.k8s.io/cluster-api/api/v1alpha3"
+	capiconditions "sigs.k8s.io/cluster-api/util/conditions"
+	ctrl "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type HandlerConfig struct {
+	CtrlClient ctrl.Client
+	Logger     micrologger.Logger
+
+	UpdateStatus      bool
+	ConditionType     capi.ConditionType
+	EnsureCreatedFunc func(ctx context.Context, object conditions.Object) error
+	EnsureDeletedFunc func(ctx context.Context, object conditions.Object) error
+}
+
+type Handler struct {
+	ctrlClient ctrl.Client
+	logger     micrologger.Logger
+
+	conditionType     capi.ConditionType
+	updateStatus      bool
+	ensureCreatedFunc func(ctx context.Context, object conditions.Object) error
+	ensureDeletedFunc func(ctx context.Context, object conditions.Object) error
+}
+
+func NewHandler(config HandlerConfig) (*Handler, error) {
+	if config.CtrlClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.CtrlClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	h := &Handler{
+		ctrlClient:        config.CtrlClient,
+		logger:            config.Logger,
+		conditionType:     config.ConditionType,
+		updateStatus:      config.UpdateStatus,
+		ensureCreatedFunc: config.EnsureCreatedFunc,
+		ensureDeletedFunc: config.EnsureDeletedFunc,
+	}
+
+	return h, nil
+}
+
+func (h *Handler) EnsureCreated(ctx context.Context, object conditions.Object) (err error) {
```

Comment:
Common stuff for all condition handlers, used by every condition handler in a decorator pattern fashion.

Reply:
This handler does the common logistics and calls into specific implementation for every condition.